03_Modules/Locations/src/module/admin.ts

```diff
@@ -0,0 +1,172 @@
+import { Service } from 'typedi';
+import { type ILogObj, Logger } from 'tslog';
+import {
+  ChargingStation,
+  Location,
+  SequelizeLocationRepository,
+} from '@citrineos/data';
+import {
+  AdminConnectorDTO,
+  AdminEVSEDTO,
+  AdminLocationDTO,
+  CitrineOcpiLocationMapper,
+  LocationDTO,
+  LocationsBroadcaster,
+  OcpiLocation,
+  OcpiLocationProps,
+  OcpiLocationRepository,
+  OcpiEvse,
+  OcpiEvseRepository,
+  OcpiConnector,
+  OcpiConnectorRepository,
+  VariableAttributesUtil,
+  OcpiLocationsUtil,
+  InvalidParamException,
+} from '@citrineos/ocpi-base';
+
+@Service()
+export class AdminLocationsService {
+  constructor(
+    private logger: Logger<ILogObj>,
+    private locationMapper: CitrineOcpiLocationMapper,
+    private locationRepository: SequelizeLocationRepository,
+    private ocpiLocationRepository: OcpiLocationRepository,
+    private ocpiEvseRepository: OcpiEvseRepository,
+    private ocpiConnectorRepository: OcpiConnectorRepository,
+    private locationsBroadcaster: LocationsBroadcaster,
+    private variableAttributesUtil: VariableAttributesUtil,
+    private ocpiLocationsUtil: OcpiLocationsUtil,
+  ) {}
+
+  public async createOrUpdateLocation(
+    adminLocationDto: AdminLocationDTO,
+    broadcast: boolean,
+  ): Promise<LocationDTO> {
+    this.logger.debug(
+      adminLocationDto.id
+        ? `Updating Location ${adminLocationDto.id}`
+        : `Creating Location ${adminLocationDto.name}`,
+    );
+
+    const [invalidLocation, message] = AdminLocationDTO.IS_LOCATION_INVALID(adminLocationDto);
+    if (invalidLocation) {
+      throw new InvalidParamException(message ?? 'Invalid location request.');
+    }
+
+    const [ocpiLocation, citrineLocation] =
+      this.mapAdminLocationDtoToEntities(adminLocationDto);
+    const savedCitrineLocation =
+      await this.locationRepository.createOrUpdateLocationWithChargingStations(
+        citrineLocation,
+      );
+    ocpiLocation[OcpiLocationProps.citrineLocationId] = savedCitrineLocation.id;
+    const savedOcpiLocation =
+      await this.ocpiLocationRepository.createOrUpdateOcpiLocation(
+        ocpiLocation,
+      );
+
+    for (const adminEvse of adminLocationDto.evses ?? []) {
+      await this.ocpiEvseRepository.createOrUpdateOcpiEvse(
+        this.mapAdminEvseDtoToOcpiEvse(adminEvse),
+      );
+      for (const adminConnector of adminEvse.connectors ?? []) {
+        await this.ocpiConnectorRepository.createOrUpdateOcpiConnector(
+          this.mapAdminConnectorToOcpiConnector(adminConnector, adminEvse),
+        );
+      }
+    }
+
+    let chargingStationVariableAttributesMap = {};
+
+    if (savedCitrineLocation.chargingPool) {
```

Comment:
So what happens if user specifies a full `AdminEvseDto` and `AdminConnectorDto` objects in the request?

I'm assuming any of the properties that are fulfilled by variable attributes are ignored?